### PR TITLE
Fixed #25172 -- Fixed check framework to work with multiple databases

### PR DESCRIPTION
**Pull request by delinhabit**
Fixes check framework handling of multiple databases, when performing field checks:
https://code.djangoproject.com/ticket/25172

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by delinhabit** [comment]
@timgraham I'll take a look at the failing tests. Hopefully I can set-up them to run locally so I can reproduce the failures.

**Comment by delinhabit** [comment]
@timgraham I fixed the python 3.4 specific issue. 

I encountered a wall with the postgres specific tests, though. ArrayField uses a subtype field (`base_field`) on which we are running the checks too. But that field will never get a `self.model` instance, and we need `self.model` in order to detect the database connection to run the checks on. Any ideas on how I can move forward with the ArrayField?

**Comment by timgraham** [comment]
I guess it might be a bug in `ArrayField` that subfield's don't get `model` set. If so, we should first fix that as a separate ticket.

**Comment by delinhabit** [comment]
Ok, Should I create a separate ticket for that and fix it first? I think I have an idea on how to fix it already.

**Comment by timgraham** [comment]
Yes please.

**Comment by delinhabit** [comment]
@timgraham I rebased with latest master (Also updated the patch in the associated Trac ticket to include the latest fixes).

**Comment by delinhabit** [comment]
Any updates on this issue? Do you have any concerns?

File: tests/check_framework/test_multi_db.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+from .tests import IsolateModelsMixin
+
+
+class TestRouter(object):
+    """
+    Dummy router that would route to the 'other' database
+    if the model name starts with 'Other'.
+    """
+    def db_for_read(self, model, **hints):
```
Comment (timgraham) [inline tests/check_framework/test_multi_db.py]: Are all the methods required? My intuition is that `allow_migrate()` would be sufficient for testing.

File: tests/check_framework/test_multi_db.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+@override_settings(DATABASE_ROUTERS=[TestRouter()])
+class TestMultiDBChecks(IsolateModelsMixin, TestCase):
+    multi_db = True
+
+    def test_checks_called_on_the_default_database(self):
+        class Model(models.Model):
+            field = models.CharField(max_length=100)
+
+        model = Model()
+        with mock.patch.object(connections['default'].validation, 'check_field') as mock_check_field:
```
Comment (timgraham) [inline tests/check_framework/test_multi_db.py]: How about also asserting the 'other' is not called?

File: tests/invalid_models_tests/test_backend_specific.py [hunk 3]
```diff
@@ -27,7 +36,7 @@ class Model(models.Model):
         # Mock connection.validation.check_field method.
         v = connection.validation
         old_check_field = v.check_field
-        v.check_field = MethodType(mock, v)
+        v.check_field = MethodType(mock_check_field, v)
```
Comment (timgraham) [inline tests/invalid_models_tests/test_backend_specific.py]: could we use `with mock.patch.object(connections['default'].validation, 'check_field')` as done in the other test? I think it's a bit cleaner than the current approach (which was done when we didn't have mock available).

File: tests/invalid_models_tests/test_backend_specific.py [hunk 4]
```diff
@@ -5,18 +5,27 @@
 
 from django.core.checks import Error
 from django.db import connection, models
+from django.test import mock
 
 from .base import IsolatedModelsTestCase
 
 
+def dummy_allow_migrate(db, app_label, **hints):
+    # Only allow_migrate on default so we can mock the check_field
```
Comment (timgraham) [inline tests/invalid_models_tests/test_backend_specific.py]: I had a hard time understanding this comment. Maybe something like `# Prevent checks from being run on the 'other' database, which doesn't have its check_field() method mocked in the test.` would be clearer.

**Comment by timgraham** [comment]
Please squash commits when updating and mark ticket as "Ready for checkin." Thanks!

**Comment by timgraham** [comment]
merged in 0cc059cd104cdb70340bd08e597d403d80dc42a6, thanks!

**Pull request closed without merging**